src/preprocessing/passes/non_clausal_simp.cpp

```diff
@@ -57,68 +80,37 @@ PreprocessingPassResult NonClausalSimp::applyInternal(
   Assert(!options::unsatCores() && !options::fewerPreprocessingHoles());
 
   d_preprocContext->spendResource(options::preprocessStep());
-
-  theory::booleans::CircuitPropagator* propagator =
-      d_preprocContext->getCircuitPropagator();
-
-  for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
-  {
-    Trace("non-clausal-simplify") << "Assertion #" << i << " : "
-                                  << (*assertionsToPreprocess)[i] << std::endl;
-  }
-
-  if (propagator->getNeedsFinish())
-  {
-    propagator->finish();
-    propagator->setNeedsFinish(false);
-  }
-  propagator->initialize();
-
-  // Assert all the assertions to the propagator
-  Trace("non-clausal-simplify") << "asserting to propagator" << std::endl;
   unsigned substs_index = d_preprocContext->getSubstitutionsIndex();
-  for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+  std::tuple<bool, std::vector<Node>> resultAndLearnedLiterals;
+  if (options::skeletonPreprocessing()){
+    // solve the boolean skeleton using cryptominisat
+    resultAndLearnedLiterals =
+        preprocessByCryptominisat(assertionsToPreprocess, substs_index);
+  }
+  else
   {
-    Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
-           == (*assertionsToPreprocess)[i]);
-    // Don't reprocess substitutions
-    if (substs_index > 0 && i == substs_index)
-    {
-      continue;
-    }
-    Trace("non-clausal-simplify")
-        << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
-    Debug("cores") << "propagator->assertTrue: " << (*assertionsToPreprocess)[i]
-                   << std::endl;
-    propagator->assertTrue((*assertionsToPreprocess)[i]);
+    // default
+    // solve the boolean skeleton using circuit propagator
+    resultAndLearnedLiterals =
+        preprocessByCircuitPropagator(assertionsToPreprocess, substs_index);
   }
 
-  Trace("non-clausal-simplify") << "propagating" << std::endl;
-  if (propagator->propagate())
+  if (!std::get<0>(resultAndLearnedLiterals))
   {
-    // If in conflict, just return false
-    Trace("non-clausal-simplify")
-        << "conflict in non-clausal propagation" << std::endl;
-    Assert(!options::unsatCores() && !options::fewerPreprocessingHoles());
-    assertionsToPreprocess->clear();
-    Node n = NodeManager::currentNM()->mkConst<bool>(false);
-    assertionsToPreprocess->push_back(n);
-    PROOF(ProofManager::currentPM()->addDependence(n, Node::null()));
-    propagator->setNeedsFinish(true);
+    handleConflictCase(assertionsToPreprocess);
```

Comment:
It looks like you call `handleConflictCase` twice when `propagator->propagate()` returns true, which does not happen in the original code. You may be able to call `handleConflictCase` only once.